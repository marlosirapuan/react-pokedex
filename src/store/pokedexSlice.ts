import { createSlice, PayloadAction } from '@reduxjs/toolkit'
import { AppThunk, AppDispatch } from 'store/store'

import { Pokemon } from 'typings'

const initialState: Pokemon[] = []

const pokedexSlice = createSlice({
  name: 'pokedex',
  initialState,
  reducers: {
    add(state, action: PayloadAction<Pokemon>) {
      state.push(action.payload)
    },
    remove(state, action: PayloadAction<string>) {
      const idx = state.findIndex((record) => record.name === action.payload)
      state.splice(idx, 1)
    }
  }
})

export const add = (pokemon: Pokemon): AppThunk => async (
  dispatch: AppDispatch
) => {
  const { name, abilities, sprites, stats, types, weight, order } = pokemon
  const newPokemon = {
    name,
    abilities,
    sprites,
    stats,
    types,
    weight,
    order
  } as Pokemon

  dispatch(pokedexSlice.actions.add(newPokemon))
}

export const remove = (name: string): AppThunk => async (
  dispatch: AppDispatch
) => {
  dispatch(pokedexSlice.actions.remove(name))
}

export default pokedexSlice.reducer
