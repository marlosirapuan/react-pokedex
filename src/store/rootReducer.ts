import { combineReducers } from '@reduxjs/toolkit'

import pokedexReducer from 'store/pokedexSlice'

const rootReducer = combineReducers({
  pokedex: pokedexReducer
})

export type RootState = ReturnType<typeof rootReducer>

export default rootReducer
