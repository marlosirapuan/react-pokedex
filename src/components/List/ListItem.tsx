import React, { useState, useEffect, useCallback } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { RootState } from 'store/rootReducer'

import { add, remove } from 'store/pokedexSlice'
import { PokemonResultItem, Pokemon } from 'typings'
import api from 'api'

import { Box, Text, Button } from '@chakra-ui/react'
import { AiOutlineStar, AiFillStar } from 'react-icons/ai'

import imageFallback from 'assets/image-not-found.png'

type ListItemProps = {
  item: PokemonResultItem
}

const ListItem = ({ item }: ListItemProps): JSX.Element => {
  const dispatch = useDispatch()
  const pokedex = useSelector((state: RootState) => state.pokedex)

  const [pokemonListItem, setPokemonListItem] = useState<Pokemon>({} as Pokemon)

  const getInfoPokemon = async (name: string): Promise<void> => {
    try {
      const { data } = await api.get<Pokemon>(`/pokemon/${name}`)
      setPokemonListItem(data)
    } catch (error) {
      setPokemonListItem({ error: true })
    }
  }

  const isFavorited = (name: string | undefined): boolean => {
    if (!name) {
      return false
    }

    return pokedex.some((pokemon) => pokemon.name === name)
  }

  const handleToggleFavorite = useCallback(
    (pokemon: Pokemon | undefined) => {
      if (pokemon) {
        if (
          pokemon.name &&
          pokedex.some((item) => item.name === pokemon.name)
        ) {
          dispatch(remove(pokemon.name))
        } else {
          dispatch(add(pokemon))
        }
      }
    },
    [dispatch, pokedex]
  )

  useEffect(() => {
    getInfoPokemon(item.name)
  }, [item.name])

  return (
    <Box
      key={pokemonListItem.name}
      display="flex"
      flexDirection="column"
      alignItems="center"
      justifyContent="center"
      minHeight={100}
      borderRadius={10}
      borderWidth={4}
      padding={3}
    >
      <Text fontSize="lg">{pokemonListItem.name}</Text>
      <img
        src={pokemonListItem.sprites?.front_default ?? imageFallback}
        alt={pokemonListItem.name}
      />
      <Button
        variant="ghost"
        _active={{ background: 'none' }}
        _hover={{ background: 'none' }}
        _focus={{ border: 'none' }}
        onClick={() => handleToggleFavorite(pokemonListItem)}
      >
        {isFavorited(pokemonListItem.name) ? (
          <AiFillStar color="orange" size={32} />
        ) : (
          <AiOutlineStar color="orange" size={32} />
        )}
      </Button>
      <Button variant="ghost">Detalhar</Button>
    </Box>
  )
}

export default ListItem
