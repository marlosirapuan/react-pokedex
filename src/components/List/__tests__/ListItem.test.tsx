import React from 'react'
import { render } from '@testing-library/react'
import { Provider } from 'react-redux'

import store from 'store/store'

import ListItem from '../ListItem'

test('renders the component', () => {
  const item = {
    name: 'Pokemon',
    url: 'http://www.image.com/image.png'
  }

  const { getByLabelText } = render(
    <Provider store={store}>
      <ListItem item={item} />
    </Provider>
  )

  const listElement = getByLabelText(/list item/i)

  expect(listElement).toBeInTheDocument()
})
