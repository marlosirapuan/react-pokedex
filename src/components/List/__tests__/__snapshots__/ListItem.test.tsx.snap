// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItem component should be able to render 1`] = `ShallowWrapper {}`;
